cmd/object-api-utils.go

```diff
@@ -675,3 +675,60 @@ func CleanMinioInternalMetadataKeys(metadata map[string]string) map[string]strin
 	}
 	return newMeta
 }
+
+// snappyCompressReader compresses data as it's read
+// from the underlying io.Reader.
+type snappyCompressReader struct {
+	r      io.Reader
+	w      *snappy.Writer
+	closed bool
+	buf    bytes.Buffer
+}
+
+func newSnappyCompressReader(r io.Reader) *snappyCompressReader {
+	cr := &snappyCompressReader{r: r}
+	cr.w = snappy.NewBufferedWriter(&cr.buf)
+	return cr
+}
+
+func (cr *snappyCompressReader) Read(p []byte) (int, error) {
+	if cr.closed {
+		// if snappy writer is closed r has been completely read,
+		// return any remaining data in buf.
+		return cr.buf.Read(p)
+	}
+
+	for {
+		// read from original using p as buffer
+		nr, readErr := cr.r.Read(p)
+
+		// write read bytes to snappy writer
+		nw, err := cr.w.Write(p[:nr])
+		if err != nil {
+			return 0, err
+		}
+		if nw != nr {
+			return 0, io.ErrShortWrite
+		}
+
+		// if last of data from reader, close snappy writer to flush
+		if readErr == io.EOF {
+			err := cr.w.Close()
+			cr.closed = true
+			if err != nil {
+				return 0, err
+			}
+		}
+
+		// read compressed bytes out of buf
+		n, err := cr.buf.Read(p)
+		if readErr != io.EOF && (err == nil || err == io.EOF) {
+			err = readErr
+		}
+
+		if n == 0 && err == nil {
```

Comment:
Can you explain a case where this condition can pass ?

Reply:
I fear this loop would never end if we intentionally give a zero sized p for the read call

Reply:
You're right about it causing an infinite loop. I've removed this check and the loop since I'm unable to reproduce the bug I thought it was correcting.